### PR TITLE
perf(indexer): optimize paged top-k-512 selection

### PR DESCRIPTION
**Pull request by voipmonitor**
## Result

Status: **implemented and qualified** for the paged sparse-attention index
selector used by GLM-5.3 prefill.

Top-k 512 uses a 1,024-entry shared candidate buffer. Top-k 1,024 and 2,048
retain an 8,192-entry buffer. The policy depends only on the compiled top-k
specialization; it is not gated by GPU architecture and does not query device
state while serving a request.

An indices-only terminal fold omits its unused FP32 score write. Every
intermediate fold still writes scores because the following fold consumes
them.

## Technical reason

The exact overflow rescan preserves top-k membership when a threshold bucket
contains more candidates than the shared candidate buffer. Reducing the buffer
for top-k 512 therefore lowers shared-memory traffic without changing selection
semantics.

Rows with fewer live candidates than `topk` initialize overflow padding before
the exact rescan. Each live index can appear at most once, and unused result
positions contain `-1`.

## Operation contract and compatibility

- Logical and physical output-index modes retain their existing semantics.
- The paged FP8 cache layout, page-table contract, supported top-k widths, and
  input layouts are unchanged.
- Score-producing callers receive the same terminal score output.
- Indices-only callers receive selected indices without a terminal score write.
- Candidate capacity remains an immutable part of the compiled kernel key.

The low-level tiled selector treats `write_values=False` as a terminal-output
contract. The paged orchestrator always enables score writes for carry-producing
launches and disables them only for the final indices-only launch.

## Correctness validation

Tests ran on an NVIDIA RTX PRO 6000 Blackwell Workstation Edition GPU with
NVIDIA compute capability 12.0 and `CUTE_DSL_ARCH=sm_120a`.

```bash
/opt/venv/bin/python -B -m pytest -q \
  tests/attention/test_benchmark_paged_indexer.py \
  tests/attention/test_attention_dsa_indexer_api.py \
  tests/attention/test_paged_prefill_topk_long_context.py \
  -k "analytic_topk or topk_candidate_capacity or row_topk or paged_prefill_topk"
```

Result: `17 passed, 29 deselected`.

The analytic benchmark oracle additionally passed:

- both 1,024- and 8,192-entry capacities at 4,080 rows and 8,192 visible
  tokens;
- a 64-token row with a one-page table and top-k 512;
- physical page-to-token arithmetic above the signed 32-bit multiplication
  boundary;
- a two-chunk 16,384-token fold, proving that intermediate carry scores remain
  usable while the terminal workspace score slice remains untouched.

The benchmark rejects `--topk-candidate-capacity` in modes that do not consume
the tiled-selector capacity policy.

## Performance qualification

The comparison used 4,080 query rows, 64 replicated indexer heads, 8,192
visible tokens, top-k 512, indices-only output, and CUDA graph replay. Both
arms used identical code, inputs, cache state, and output semantics; only the
compile-time candidate capacity differed.

| Candidate capacity | Median latency | Relative selector throughput |
|---:|---:|---:|
| 1,024 | 961.54 us | 1.0243x |
| 8,192 | 984.86 us | 1.0000x |

Both kernels use 27 registers per thread, zero stack bytes, zero local-memory
bytes, and one 1,024-thread cooperative thread array per streaming
multiprocessor. Dynamic shared memory is 20,096 bytes for 1,024 candidates and
77,440 bytes for 8,192 candidates. The smaller policy improves selector
throughput by 2.43% without changing occupancy, register use, or spill behavior.

The commands, source revisions, hardware identity, raw graph-replay samples,
balanced process ordering, and compiled-resource evidence are recorded in
`docs/evidence/paged_topk_sm120_prefill.md`.

This evidence qualifies the selector operation. It does not infer end-to-end
model throughput from one kernel family.

## AI assistance disclosure

AI assistance was used to implement and validate the selector changes and to
prepare this pull-request description.

## Summary

- Selects shared candidate capacity by compile-time `topk`: 1,024 entries for top-k 512 and 8,192 entries for top-k 1,024 and 2,048.
- Skips terminal FP32 score writes for indices-only requests. Intermediate fold scores remain available for later folds.
- Preserves existing output modes, cache layout, page-table contract, input layouts, supported top-k widths, and score-producing behavior.
- Adds overflow and padding handling that produces distinct, valid candidates before rescanning.
- Adds bounded page-table validation and 64-bit physical-slot arithmetic.
- Adds correctness coverage for candidate capacities, indices-only output, CUDA graph replay, tie filling, long-context rows, and large page IDs.
- Validation passed 17 tests, with 29 deselected. Benchmarking measured 961.54 µs for 1,024 candidates versus 984.86 µs for 8,192 candidates, or 1.0243× relative selector throughput.
- Adds reproducible SM120 qualification evidence, including latency, resource measurements, oracle cases, and remaining end-to-end scope.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by voipmonitor** [comment]
@coderabbitai review

File: b12x/attention/dsa_indexer/tiled_topk.py [hunk 1]
```diff
@@ -1494,6 +1544,10 @@ def run_tiled_topk(
     flat_carry_values = carry_values.reshape(-1).contiguous()
     flat_carry_indices = carry_indices.reshape(-1).contiguous()
 
+    smem_candidate_capacity = _resolve_smem_candidate_capacity(
+        topk=topk,
+        device=tile_logits.device,
+    )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply b12x/attention/dsa_indexer/tiled_topk.py]: Addressed in ee611d20. Shared-memory candidate capacity is cached by resolved CUDA device index and top-k, and cache reset is integrated with the selector cache lifecycle. The unit selection verifies one capability query per immutable key; it passed 7 tests, and the selector GPU suite passed 15 tests.

File: docs/evidence/paged_topk_sm120_prefill.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  tests/attention/test_paged_prefill_topk_long_context.py \
+  -k "row_topk or paged_prefill_topk"
+```
+
+Result: `15 passed, 29 deselected`.
+
+A production-geometry analytic run used 4,080 query rows, 64 replicated
+indexer heads, an 8,192-token visible cache, top-k 512, and a shared paged
+table. Every selected index matched the exact FP32 top-k membership and the
+maximum valid-score error was 0.000488.
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply docs/evidence/paged_topk_sm120_prefill.md]: Addressed in b68c1972. The evidence now records exact worktrees, explicit physical GPU and sm_120a selection, P1 clocks/throttle state, CUTLASS DSL/PTXAS versions, balanced arm ordering, source-specific cache-manifest hashes, cold/preconditioned measurements, 30 raw replay samples per arm, and separate analytic and indices-only correctness coverage.

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by voipmonitor** [comment]
@coderabbitai full review

File: b12x/attention/dsa_indexer/tiled_topk.py [hunk 3]
```diff
@@ -47,17 +47,41 @@
...
 
 
+@lru_cache(maxsize=32)
+def _resolve_smem_candidate_capacity_for_device(
+    topk: int,
+    device_index: int,
+) -> int:
+    """Resolve immutable selector geometry once per device and top-k."""
+    capability = torch.cuda.get_device_capability(device_index)
+    if capability == (12, 0) and int(topk) == 512:
```
Comment (lukealonso) [inline b12x/attention/dsa_indexer/tiled_topk.py]: Why is this SM120-only? If 1024 avoids an SM120-specific compiler/resource cliff, please include identical-semantics 1024-vs-8192 timings plus PTXAS register, spill, local-memory, shared-memory, and occupancy evidence. Otherwise make this a top-k-only policy and remove the capability branch.

**Review by lukealonso — CHANGES_REQUESTED** [review]
<sub>Requested changes in [CodeRabbit Change Stack](https://app.coderabbit.ai/change-stack/local-inference-lab/b12x/pull/260)</sub>

File: b12x/attention/dsa_indexer/tiled_topk.py (hunk 3 above)
Comment (voipmonitor) [reply b12x/attention/dsa_indexer/tiled_topk.py]: The candidate-capacity policy is now architecture-neutral and depends only on the compile-time top-k width. Top-k 512 uses 1,024 entries; top-k 1,024 and 2,048 retain 8,192 entries. An identical-semantics CUDA-graph comparison measured 1,030.14 us for 1,024 entries and 1,044.48 us for 8,192 entries. Both kernels use 27 registers/thread, zero stack/local memory, and one CTA/SM; dynamic shared memory is 20,096 versus 77,440 bytes. The focused selector suite reports 16 passed and 29 deselected. The PR body and repository evidence now state the operation contract, resource data, raw timing provenance, and scope.

**Comment by voipmonitor** [comment]
@coderabbitai full review

File: benchmarks/benchmark_paged_indexer.py [hunk 4]
```diff
@@ -350,6 +371,29 @@ def main() -> None:
     parser.add_argument("--seed", type=int, default=91_100)
     args = parser.parse_args()
 
+    if args.check and args.indices_only:
+        raise ValueError("--check requires score output and cannot use --indices-only")
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/benchmark_paged_indexer.py]: Addressed in 03ba9855. The benchmark now permits `--check --indices-only`, validates every selected index against the analytic top-k oracle, and verifies that a preinitialized caller-owned score buffer remains unchanged before timing. The exact 4,080-row production geometry passed for both 1,024- and 8,192-entry capacities, and the focused GPU selector suite passed 16 tests with 29 deselected.

File: docs/evidence/paged_topk_sm120_prefill.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+Result: `16 passed, 29 deselected`. The selection covers top-k policy,
+indices-only output, short-row padding, exact overflow selection, and CUDA graph
+replay with changing live inputs.
+
+## Selector performance
+
+| Candidate capacity | Process medians, microseconds | Median |
+| ---: | --- | ---: |
+| 1,024 | 1029.38, 1029.73, 1030.16, 1030.14, 1031.10 | 1030.14 us |
+| 8,192 | 1044.19, 1045.50, 1044.48, 1045.74, 1044.48 | 1044.48 us |
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply docs/evidence/paged_topk_sm120_prefill.md]: Addressed in d5167b8a. The evidence now records both preconditioning processes, the balanced A/B process order, and all 30 graph-replay samples for every measured process. The oracle-validated process medians are 961.54 us for 1,024 entries and 984.86 us for 8,192 entries, which is 2.43% higher selector throughput in the stated ratio direction.

**Comment by voipmonitor** [comment]
@coderabbitai full review

**Comment by voipmonitor** [comment]
@lukealonso The architecture-specific capacity gate is removed. Candidate
capacity is now a compile-time top-k policy: top-k 512 uses 1,024 entries, while
top-k 1,024 and 2,048 retain 8,192 entries.

The identical-semantics comparison measures 961.54 us for 1,024 candidates and
984.86 us for 8,192 candidates, or 2.43% higher selector throughput. Both
specializations use 27 registers per thread, zero stack or local memory, and one
thread block per streaming multiprocessor; dynamic shared memory is 20,096
versus 77,440 bytes.

The remaining review findings are also addressed:

- indices-only validation now checks the terminal workspace score slice that
  the paged selector actually receives;
- short-row oracle page columns are bounded before gathering;
- capacity overrides are rejected for benchmark modes that do not consume
  them;
- a two-chunk analytic case proves that carry-producing folds retain their
  score values while the terminal indices-only slice remains untouched;
- 16 focused selector tests pass.

The pull-request description and repository evidence contain the operation
contract, commands, raw samples, and resource data.

@coderabbitai full review

**Comment by voipmonitor** [comment]
The remaining oracle and provenance findings are addressed.

- Page-based token-slot arithmetic now widens page IDs and offsets to 64 bits
  before multiplication and addition.
- A CPU oracle test covers a page ID above the signed 32-bit multiplication
  boundary.
- The focused suite reports 17 passed and 29 deselected.
- The evidence maps both timed selector objects to their source revision,
  package fingerprint, compile-manifest hash and cache key, compile-specification
  hash, CUTLASS DSL version, and PTXAS executable identity.
- The 4,080-row production-shaped indices-only oracle still passes after the
  change.

@coderabbitai full review

**Comment by voipmonitor** [comment]
@coderabbitai review

The indices-only evidence now states the implemented validation contract directly: the benchmark initializes the binding-owned terminal workspace score slice and verifies that `index_topk_fp8` leaves it unchanged. Please review the pull-request head.

**Comment by voipmonitor** [comment]
@lukealonso The SM120-only policy from your review is no longer present. The current branch makes the 1,024-entry candidate buffer a top-k-512 policy, preserves 8,192 entries for top-k 1,024 and 2,048, and includes identical-semantics resource/timing evidence plus the indices-only and large-page-ID correctness oracles. Could you re-review the current head? The complete remaining merge list is summarized in local-inference-lab/vllm#548.

**Comment by voipmonitor** [comment]
Superseded by #316. The replacement applies the reviewed architecture-independent top-k-512 candidate-capacity policy to B12X master at 613cfd9b, preserves the terminal indices-only and overflow-correctness changes, and retains the matched 1,024-versus-8,192 timing and compiled-resource evidence requested in review. This branch no longer merges with master, so it is closed to keep one mergeable implementation.
